polkadot/node/core/approval-voting/src/criteria.rs

```diff
@@ -178,29 +174,30 @@ fn relay_vrf_modulo_cores(
 			"Suboptimal configuration `num_samples` should be less than `n_cores` / 2",
 		);
 	}
-
-	let mut rand_chacha =
+	let rand_chacha =
 		ChaCha20Rng::from_seed(vrf_in_out.make_bytes::<<ChaCha20Rng as SeedableRng>::Seed>(
 			approval_types::v2::CORE_RANDOMNESS_CONTEXT,
 		));
+	generate_samples(rand_chacha, num_samples as usize, max_cores as usize)
+}
 
-	let mut res = HashSet::with_capacity(num_samples as usize);
-	while res.len() < min(num_samples, max_cores) as usize {
-		// Production networks should run with num_samples no larger than `n_cores/2`, so on worst
-		// case on each iteration this loop has >1/2 chance of finishing.
-		// With our current production parameters the chances for each iteration not hitting
-		// a duplicate are around 90%.
-		// See: https://github.com/paritytech/polkadot-sdk/pull/1178#discussion_r1376501206
-		// for more details.
-		for _ in 0..MAX_SAMPLING_ITERATIONS {
-			let random_core = rand_chacha.gen_range(0..max_cores).into();
-			if !res.contains(&random_core) {
-				res.insert(random_core);
-				break
-			}
-		}
-	}
-	res.into_iter().collect_vec()
+/// Generates `num_sumples` randomly from (0..max_cores) range
+///
+/// Note! The algorithm can't change because validators on the other
+/// side won't be able to check the assignments until they update.
+/// This invariant is tested with `generate_samples_invariant`, so the
+/// tests will catch any subtle changes in the implementation of this function
+/// and its dependencies.
+fn generate_samples(
+	mut rand_chacha: ChaCha20Rng,
+	num_samples: usize,
+	max_cores: usize,
+) -> Vec<CoreIndex> {
+	let num_samples = max(1, min(num_samples, max_cores / 2));
+
+	let mut random_cores = (0..max_cores as u32).map(|val| val.into()).collect::<Vec<CoreIndex>>();
+	let (samples, _) = random_cores.partial_shuffle(&mut rand_chacha, num_samples as usize);
+	samples.into_iter().map(|val| *val).collect_vec()
```

Comment:
Apologies I confused you.. 

Alone `let num_samples = max(1, min(num_samples, max_cores / 2));` would've fixed the previous code.

This code feels less efficient, but afaik should be fine in practice, and does not require the `let num_samples = ` change.